File: config-biz/src/main/java/io/gitee/zicai/biz/service/BaseService.java
package io.gitee.zicai.biz.service;

import com.github.pagehelper.PageHelper;
import com.github.pagehelper.PageInfo;
import io.gitee.zicai.core.entity.BaseEntity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;
import tk.mybatis.mapper.common.Mapper;

public abstract class BaseService<M extends Mapper<T>, T extends BaseEntity> {

    @Autowired
    protected M mapper;

    public T selectByPrimaryKey(Long id) {
        return mapper.selectByPrimaryKey(id);
    }

    @Transactional
    public int insert(T entity) {
        return mapper.insertSelective(entity);
    }

    @Transactional
    public int deleteByPrimaryKey(long id) {
        return mapper.deleteByPrimaryKey(id);
    }

    @Transactional
    public int deleteExample(Object example) {
        return mapper.deleteByExample(example);
    }

    @Transactional
    public int updateByPrimaryKey(T entity) {
        return mapper.updateByPrimaryKeySelective(entity);
    }

    @Transactional
    public int updateByExample(T entity, Object example) {
        return mapper.updateByExampleSelective(entity, example);
    }

    public PageInfo<T> page(int index, int size) {
        return page(index, size, null);
    }

    public PageInfo<T> page(int index, int size, Object example) {
        PageHelper.startPage(index, size);
        PageHelper.orderBy("create_time DESC");
        return new PageInfo<>(mapper.selectByExample(example));
    }
}

File: config-biz/src/main/java/io/gitee/zicai/biz/service/BaseService.kt
package io.gitee.zicai.biz.service

import com.baomidou.mybatisplus.core.conditions.Wrapper
import com.baomidou.mybatisplus.core.mapper.BaseMapper
import com.baomidou.mybatisplus.core.metadata.IPage
import com.baomidou.mybatisplus.extension.plugins.pagination.Page
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import io.gitee.zicai.core.entity.BaseEntity

abstract class BaseService<M : BaseMapper<T>, T : BaseEntity> : ServiceImpl<M, T>() {

    fun selectByPrimaryKey(id: Long?): T {
        return this.getById(id)
    }

    fun insert(entity: T): Int {
        return baseMapper.insert(entity)
    }

    fun deleteByPrimaryKey(id: Long): Int {
        return baseMapper.deleteById(id)
    }

    fun deleteExample(wrapper: Wrapper<T>): Int {
        return baseMapper.delete(wrapper)
    }

    fun updateByPrimaryKey(entity: T): Int {
        return baseMapper.updateById(entity)
    }

    fun updateByExample(entity: T, wrapper: Wrapper<T>): Int {
        return baseMapper.update(entity, wrapper)
    }

    @JvmOverloads
    fun page(index: Int, size: Int, wrapper: Wrapper<T>? = null): IPage<T> {
        val page = Page<T>(index.toLong(), size.toLong())
        page.setDesc("create_time")
        return baseMapper.selectPage(page, wrapper)
    }
}
